src/zcash/JoinSplit.cpp

```diff
@@ -0,0 +1,344 @@
+                    phi,
+                    rt,
+                    h_sig,
+                    inputs,
+                    output_notes,
+                    vpub_old,
+                    vpub_new
+                );
+            }
+
+            if (!pb.is_satisfied()) {
+                throw std::invalid_argument("Constraint system not satisfied by inputs");
+            }
+
+            primary_input = pb.primary_input();
+            aux_input = pb.auxiliary_input();
+        }
+
+        auto proof = r1cs_ppzksnark_prover<ppzksnark_ppT>(
+            *pk,
+            primary_input,
+            aux_input
+        );
+
+        std::stringstream ss;
+        ss << proof;
+
+        return ss.str();
+    }
+};
+
+template<size_t NumInputs, size_t NumOutputs>
+JoinSplit<NumInputs, NumOutputs>* JoinSplit<NumInputs, NumOutputs>::Generate()
+{
+    JoinSplitCircuit<NumInputs, NumOutputs>::initialize();
+    auto js = new JoinSplitCircuit<NumInputs, NumOutputs>();
+    js->generate();
+
+    return js;
+}
+
+template<size_t NumInputs, size_t NumOutputs>
+JoinSplit<NumInputs, NumOutputs>* JoinSplit<NumInputs, NumOutputs>::Unopened()
+{
+    JoinSplitCircuit<NumInputs, NumOutputs>::initialize();
+    return new JoinSplitCircuit<NumInputs, NumOutputs>();
+}
+
+template<size_t NumInputs, size_t NumOutputs>
+uint256 JoinSplit<NumInputs, NumOutputs>::h_sig(
+    const uint256& randomSeed,
+    const boost::array<uint256, NumInputs>& nullifiers,
+    const uint256& pubKeyHash
+) {
+    unsigned char personalization[crypto_generichash_blake2b_PERSONALBYTES]
+        = {'Z','c','a','s','h','C','o','m','p','u','t','e','h','S','i','g'};
+
+    std::vector<unsigned char> block(randomSeed.begin(), randomSeed.end());
+
+    for (size_t i = 0; i < NumInputs; i++) {
+        block.insert(block.end(), nullifiers[i].begin(), nullifiers[i].end());
+    }
+
+    block.insert(block.end(), pubKeyHash.begin(), pubKeyHash.end());
+
+    uint256 output;
+
+    if (crypto_generichash_blake2b_salt_personal(output.begin(), 32,
+                                                 &block[0], block.size(),
+                                                 NULL, 0, // No key.
+                                                 NULL,    // No salt.
+                                                 personalization
+                                                ) != 0)
+    {
+        throw std::logic_error("hash function failure");
+    }
+
+    return output;
+}
+
+Note JSOutput::note(const uint256& phi, const uint256& r, size_t i, const uint256& h_sig) const {
+    uint256 rho = PRF_rho(phi, i, h_sig);
+
+    return Note(addr.a_pk, value, rho, r);
+}
+
+JSOutput::JSOutput() : addr(uint256(), uint256()), value(0) {
+    SpendingKey a_sk(random_uint256());
+    addr = a_sk.address();
+}
+
+JSInput::JSInput() : witness(ZCIncrementalMerkleTree().witness()),
+                     key(random_uint256()) {
+    note = Note(key.address().a_pk, 0, random_uint256(), random_uint256());
```

Comment:
It's not strictly necessary to generate a_pk from a_sk for a dummy input; it could just be a random uint256 (which is indistinguishable from a PRF_addr output). The commitment does have to be consistent with the chosen a_pk though, so perhaps it's easier to do it this way for now. Incidentally, I think `cm()` is recomputing a_pk which it shouldn't need to do. Maybe we should cache it, but I'm not sure it's significant for performance anyway.